src/lib/autoinstall/ask/runner.rb

```diff
@@ -0,0 +1,215 @@
+# Copyright (c) [2021] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "autoinstall/widgets/ask_dialog"
+require "autoinstall/ask/profile_reader"
+require "autoinstall/autoinst_profile/ask_list_section"
+require "autoinstall/script_runner"
+
+Yast.import "Profile"
+
+module Y2Autoinstall
+  module Ask
+    # Asks the questions defined in the <ask-list> section of an AutoYaST profile
+    #
+    # ## How does it work?
+    #
+    # The process starts by grouping all the questions defined in the <ask-list> section
+    # by their `dialog` ID. All the questions with the same ID will be placed in the same
+    # dialog. If an ID is not specified, the question is shown on a separate dialog.
+    #
+    # Then the first dialog is presented to the user. When she/he presses the 'Next'
+    # button, this runner processes the answers and goes to the next dialog. Processing
+    # the response implies:
+    #
+    # * Updating the profile if an AutoYaST path was given (elements `path` and
+    #   `pathlist` from the profile).
+    # * Writing the answer to the file, if a `file` element was specified.
+    # * Running the given script (`script` element from the profile).
+    #
+    # @example Ask questions from the profile during the 1st stage
+    #   Yast::Profile.ReadXML("autoinst.xml")
+    #   runner = Runner.new(profile, stage: :initial)
+    #   runner.run
+    class Runner
+      # @return [Hash] AutoYaST profile
+      attr_reader :profile
+      # @return [Symbol] Stage to run
+      attr_reader :stage
+
+      # @param profile [Hash] AutoYaST profile
+      # @param stage [Symbol] :initial or :cont
```

Comment:
I see stage here so many times and with so big chance for typo, that I prefer if it can be constants somewhere defined and instead constants used here. or maybe even Stage object ( which kind of seems like overkill )